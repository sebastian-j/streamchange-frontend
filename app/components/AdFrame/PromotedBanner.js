import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { FormattedMessage } from 'react-intl';
import messages from './messages';

const AdTitle = styled.span`
  color: #7b7b7b;
  font-family: Arial, sans-serif;
  margin-bottom: 5px;
  margin-right: ${({ margin }) => margin}px;
`;

const PromotedContentWrapper = styled.div`
  margin-bottom: ${({ margin }) => margin}px;
  position: relative;
  top: calc(45vh - ${({ offset }) => offset}px);
`;

const Image = styled.img`
  width: 31vw;
`;

const Shield = styled.div`
  bottom: 0;
  left: 0;
  padding-bottom: ${({ margin }) => margin}px;
  position: absolute;
  right: 0;
  top: 10px;
`;

const PromotedBanner = (props) => {
  const [imgHeight, setImgHeight] = useState(0);
  const isVideo = props.imageUrl.substr(props.imageUrl.length - 3) === 'mp4';

  const onImgLoad = ({ target: img }) => {
    setImgHeight(img.offsetHeight);
  };

  const onVideoLoad = ({ target: video }) => {
    setImgHeight(Math.round(video.videoHeight * 0.6));
  };

  return (
    <PromotedContentWrapper
      offset={imgHeight}
      margin={props.testMargins ? 1 : Math.round(Math.random() * 1000) / 100}
    >
      <AdTitle
        margin={props.testMargins ? 1 : Math.round(Math.random() * 1000) / 100}
      >
        <FormattedMessage {...messages.title} />
      </AdTitle>
      <a href={props.channelUrl} target="_blank">
        <div>
          {!isVideo && (
            <Image alt={props.title} onLoad={onImgLoad} src={props.imageUrl} />
          )}
          {isVideo && (
            <video
              width="100%"
              title={props.title}
              autoPlay
              loop
              muted
              onLoadedMetadata={onVideoLoad}
            >
              <source src={props.imageUrl} type="video/mp4" />
            </video>
          )}
        </div>
        <Shield
          margin={
            props.testMargins ? 1 : Math.round(Math.random() * 1000) / 100
          }
        />
        <Shield
          margin={
            props.testMargins ? 1 : Math.round(Math.random() * 1000) / 100
          }
        />
        <Shield
          margin={
            props.testMargins ? 1 : Math.round(Math.random() * 1000) / 100
          }
        />
      </a>
    </PromotedContentWrapper>
  );
};

PromotedBanner.propTypes = {
  channelUrl: PropTypes.string.isRequired,
  imageUrl: PropTypes.string.isRequired,
  testMargins: PropTypes.bool,
  title: PropTypes.string,
};

PromotedBanner.defaultProps = {
  testMargins: false,
};

export default PromotedBanner;
