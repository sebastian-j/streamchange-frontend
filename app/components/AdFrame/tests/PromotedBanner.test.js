import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import PromotedBanner from '../PromotedBanner';

const renderer = new ShallowRenderer();

describe('<PromotedBanner />', () => {
  it('should render and match the snapshot', () => {
    renderer.render(
      <PromotedBanner
        channelUrl="https://www.youtube.com/channel/UC3GumCi7taJQ0wWbKK-hR2w"
        imageUrl="url"
        testMargins
        title="test-advertisement"
      />,
    );
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
