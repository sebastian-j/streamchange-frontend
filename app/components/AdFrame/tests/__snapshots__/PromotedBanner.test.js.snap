// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PromotedBanner /> should render and match the snapshot 1`] = `
<PromotedBanner__PromotedContentWrapper
  margin={1}
  offset={0}
>
  <PromotedBanner__AdTitle
    margin={1}
  >
    <FormattedMessage
      defaultMessage="Advertisement"
      id="streamchange.components.AdFrame.title"
      values={Object {}}
    />
  </PromotedBanner__AdTitle>
  <a
    href="https://www.youtube.com/channel/UC3GumCi7taJQ0wWbKK-hR2w"
    target="_blank"
  >
    <div>
      <PromotedBanner__Image
        alt="test-advertisement"
        onLoad={[Function]}
        src="url"
      />
    </div>
    <PromotedBanner__Shield
      margin={1}
    />
    <PromotedBanner__Shield
      margin={1}
    />
    <PromotedBanner__Shield
      margin={1}
    />
  </a>
</PromotedBanner__PromotedContentWrapper>
`;
